File: 42-trapping-rain-water/trapping-rain-water.py
class Solution(object):
    def trap(self, bars):
        """
        :type height: List[int]
        :rtype: int
        """
        left = 0
        right = len(bars) - 1
        minimum = 0
        maximum = 0
        result = 0
        while (left < right):
            minimum = min(bars[left], bars[right])
            maximum = max(maximum, minimum)
            result += maximum - minimum
            if bars[left] < bars[right]:
                left += 1
            else: 
                right -= 1
        return result

File: 42-trapping-rain-water/trapping-rain-water.java
class Solution {
    public int trap(int[] bars) {
        int left = 0;
        int right = bars.length - 1;
        int result = 0;
        int max = 0;
        int min = 0;
        while (left < right) {
            min = Math.min(bars[left], bars[right]);
            max = Math.max(max, min);
            result += max - min;
            if (bars[left] < bars[right]) {
                left++;
            }
            else {
                right--;
            }
        }
        return result; 
    }
}